### PR TITLE
[subdevice] items_optional for Dimmable light

### PR DESCRIPTION
**Pull request by Zehir**
Part 10 of https://github.com/dresden-elektronik/deconz-rest-plugin/pull/7155

### CODE REVIEW

**Comment by manup** [comment]
Ha interesting, @ebaauw if we have a light which doesn't support `cap/on/off_with_effect` is it better to not have the item and implicitly treat it as `false`, or have it and explicitly set it to `false`?

**Comment by ebaauw** [comment]
My idea was to include `capabilities` items only when the light supports the corresponding capability, similar to only including supported `config` items.  Not sure if that’s best, but it seems simpler and maybe cleaner to me.  On the other hand, making these mandatory would force people to think about them when creating a DDF.

Note that the `color` `capabilities` and `config` items don’t apply to a dimmable light, and shouldn’t even be optional.  DDF should overwrite the device type (to color temperature light or (extended) color light) when the device does support these.  Also the old ZHA color dimmable light should not be used in a DDF, in favour of (extended) color light.

**Comment by Zehir** [comment]
Maybe we should add a third array with a list of forbidden attributes ?

**Comment by ebaauw** [comment]
> Maybe we should add a third array with a list of forbidden attributes ?

That’s redundant if you have optional attributes.

**Comment by Zehir** [comment]
> 

Or maybe a rule that say if there is the attribute `capabilities/color` you should not use type `$TYPE_DIMMABLE_LIGHT`

**Comment by ebaauw** [comment]
Huh?  Either we specify mandatory and optional items, implying the rest are not allowed, or we specify mandatory and forbidden item, implying the rest are optional.  I don't see why we need to specify mandatory, optional, and forbidden.

Either we specify the items per device type, or the device types per item.  I don't see why we want to specify mandatory items per device type, but forbidden devices types per item.

File: devices/generic/subdevices/dimmable_light.json [hunk 1]
```diff
@@ -6,14 +6,21 @@
     "order": 11,
     "uuid": ["$address.ext", "0x01"],
     "items": [
```
Comment (ebaauw) [inline devices/generic/subdevices/dimmable_light.json:8]: `state/reachable` should be mandatory.

File: devices/generic/subdevices/dimmable_light.json [hunk 2]
```diff
@@ -6,14 +6,21 @@
...
+        "state/alert",
+        "state/bri",
+        "state/on"
+    ],
+    "items_optional":[
         "cap/alert/trigger_effect",
         "cap/bri/move_with_onoff",
+        "cap/color/capabilities",
```
Comment (ebaauw) [inline devices/generic/subdevices/dimmable_light.json:16]: Not for _Dimmable Light_.

File: devices/generic/subdevices/dimmable_light.json [hunk 3]
```diff
@@ -6,14 +6,21 @@
...
+        "cap/groups_not_supported",
         "cap/on/off_with_effect",
+        "cap/transition_block",
         "config/bri/execute_if_off",
         "config/bri/startup",
-        "config/on/startup",
-        "state/alert",
-        "state/bri",
-        "state/on"
+        "config/color/ct/startup",
```
Comment (ebaauw) [inline devices/generic/subdevices/dimmable_light.json:22]: Not for _Dimmable Light_.

File: devices/generic/subdevices/dimmable_light.json [hunk 4]
```diff
@@ -6,14 +6,21 @@
...
+        "config/color/execute_if_off",
```
Comment (ebaauw) [inline devices/generic/subdevices/dimmable_light.json:23]: Not for _Dimmable Light_.

File: devices/generic/subdevices/dimmable_light.json [hunk 5]
```diff
@@ -6,14 +6,21 @@
...
+        "state/bri",
+        "state/on"
+    ],
+    "items_optional":[
         "cap/alert/trigger_effect",
         "cap/bri/move_with_onoff",
+        "cap/color/capabilities",
+        "cap/groups_not_supported",
         "cap/on/off_with_effect",
+        "cap/transition_block",
```
Comment (ebaauw) [inline devices/generic/subdevices/dimmable_light.json:19]: Not for _Dimmable Light_.

**Review by ebaauw — CHANGES_REQUESTED** [review]
See also https://github.com/dresden-elektronik/deconz-rest-plugin/wiki/Mandatory-and-Optional-Items-for-Lights-Resources.

File: devices/generic/subdevices/dimmable_light.json [hunk 6]
```diff
@@ -6,14 +6,21 @@
...
     "items": [
+        "state/alert",
+        "state/bri",
+        "state/on"
+    ],
+    "items_optional":[
         "cap/alert/trigger_effect",
         "cap/bri/move_with_onoff",
+        "cap/color/capabilities",
+        "cap/groups_not_supported",
```
Comment (ebaauw) [inline devices/generic/subdevices/dimmable_light.json:17]: `cap/groups/not_supported`

**Review by ebaauw — CHANGES_REQUESTED** [review]

File: devices/generic/subdevices/dimmable_light.json (hunk 1 above)
Comment (Zehir) [reply devices/generic/subdevices/dimmable_light.json:8]: Done

File: devices/generic/subdevices/dimmable_light.json (hunk 2 above)
Comment (Zehir) [reply devices/generic/subdevices/dimmable_light.json:16]: Removed

File: devices/generic/subdevices/dimmable_light.json (hunk 3 above)
Comment (Zehir) [reply devices/generic/subdevices/dimmable_light.json:22]: Removed

File: devices/generic/subdevices/dimmable_light.json (hunk 4 above)
Comment (Zehir) [reply devices/generic/subdevices/dimmable_light.json:23]: Removed

File: devices/generic/subdevices/dimmable_light.json (hunk 5 above)
Comment (Zehir) [reply devices/generic/subdevices/dimmable_light.json:19]: Removed

File: devices/generic/subdevices/dimmable_light.json (hunk 6 above)
Comment (Zehir) [reply devices/generic/subdevices/dimmable_light.json:17]: Fixed

**Review by ebaauw — APPROVED** [review]

**Pull request merged**